File: Public/Proxy/Routers/subtitle.py
# Bu araç @keyiflerolsun tarafından | @KekikAkademi için yazılmıştır.

from fastapi        import Request, Response
from .              import proxy_router
from ..Libs.helpers import prepare_request_headers, process_subtitle_content, CORS_HEADERS
from urllib.parse   import unquote
import httpx

@proxy_router.get("/subtitle")
async def subtitle_proxy(request: Request, url: str, referer: str = None, user_agent: str = None):
    """Altyazı proxy endpoint'i"""
    try:
        decoded_url     = unquote(url)
        request_headers = prepare_request_headers(request, decoded_url, referer, user_agent)
        
        async with httpx.AsyncClient(follow_redirects=True, timeout=30.0, verify=False) as client:
            response = await client.get(decoded_url, headers=request_headers)
            
            if response.status_code >= 400:
                return Response(
                    content     = f"Altyazı hatası: {response.status_code}", 
                    status_code = response.status_code
                )
            
            processed_content = process_subtitle_content(
                response.content, 
                response.headers.get("content-type", ""), 
                decoded_url
            )
            
            return Response(
                content     = processed_content,
                status_code = 200,
                headers     = {"Content-Type": "text/vtt; charset=utf-8", **CORS_HEADERS},
                media_type  = "text/vtt"
            )
            
    except Exception as e:
        return Response(
            content     = f"Proxy hatası: {str(e)}", 
            status_code = 500
        )

File: services/proxy/handlers/subtitle.go
// Bu araç @keyiflerolsun tarafından | @KekikAkademi için yazılmıştır.

package handlers

import (
	"io"
	"kekik-proxy/utils"
	"net/http"
	"net/url"

	"github.com/gin-gonic/gin"
	"github.com/pterm/pterm"
)

// SubtitleProxy altyazı proxy endpoint'i
func SubtitleProxy(c *gin.Context) {
	urlParam := c.Query("url")
	referer := c.Query("referer")
	userAgent := c.Query("user_agent")

	if urlParam == "" {
		c.String(http.StatusBadRequest, "URL parametresi gerekli")
		return
	}

	decodedURL, err := url.QueryUnescape(urlParam)
	if err != nil {
		c.String(http.StatusBadRequest, "Geçersiz URL")
		return
	}

	// Request headers hazırla
	reqHeaders := utils.PrepareRequestHeaders(referer, userAgent)

	// Upstream request oluştur
	req2, err := http.NewRequest("GET", decodedURL, nil)
	if err != nil {
		c.String(http.StatusBadGateway, "Request oluşturulamadı: "+err.Error())
		return
	}

	// Headers kopyala
	for k, v := range reqHeaders {
		req2.Header[k] = v
	}

	// Upstream'e istek yap
	resp, err := httpClient.Do(req2)
	if err != nil {
		pterm.Error.Printf("Upstream Error (Subtitle): %s -> %v\n", decodedURL, err)
		c.String(http.StatusBadGateway, "Proxy Error: "+err.Error())
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		c.String(resp.StatusCode, "Altyazı hatası: "+resp.Status)
		return
	}

	// Body oku
	content, err := io.ReadAll(resp.Body)
	if err != nil {
		c.String(http.StatusBadGateway, "Body okuma hatası")
		return
	}

	// İçeriği işle (SRT -> VTT dönüşümü vs.)
	contentType := resp.Header.Get("Content-Type")
	processedContent := utils.ProcessSubtitleContent(content, contentType, decodedURL)

	// Response headers
	c.Header("Content-Type", "text/vtt; charset=utf-8")
	c.Header("Access-Control-Allow-Origin", "*")
	c.Header("Access-Control-Allow-Methods", "GET, HEAD, OPTIONS")
	c.Header("Access-Control-Allow-Headers", "Origin, Content-Type, Accept, Range")

	c.Data(http.StatusOK, "text/vtt", processedContent)
}
